### PR TITLE
[1.10.x] Properly encode model and dataclass default for schema

### LINKED ISSUE #4774: `schema_json` can't encode default value of`dict[UUID, Any]` and raises `TypeError`
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When a model has default value of `{uuid4(): "..."}`, and being a dafault value of another model, `json_schema` fails to generate schema with an error.

### Example Code

```Python
from uuid import uuid4

from pydantic import BaseModel
from pydantic import UUID4


class Service(BaseModel):
    urls: dict[UUID4, str] = {uuid4(): "https://example.com"}


class Settings(BaseModel):
    service: Service = Service()


print(Settings.schema_json(indent=4, ensure_ascii=False))
```

```py
Traceback (most recent call last):
  File "/Users/bobronium/Library/Application Support/JetBrains/PyCharm2022.3/scratches/scratch_131.py", line 15, in <module>
    print(Settings.schema_json(indent=4, ensure_ascii=False))
  File "pydantic/main.py", line 674, in pydantic.main.BaseModel.schema_json
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  [Previous line repeated 1 more time]
  File "/Users/bobronium/.pyenv/versions/3.10.1/lib/python3.10/json/encoder.py", line 376, in _iterencode_dict
    raise TypeError(f'keys must be str, int, float, bool or None, '
TypeError: keys must be str, int, float, bool or None, not UUID
```

### This works, though

```py
from uuid import uuid4

from pydantic import BaseModel
from pydantic import UUID4


class Service(BaseModel):
    urls: dict[UUID4, str] = {uuid4(): "https://example.com"}


class Settings(BaseModel):
    service: Service  # <--------- Notice no default value here


print(Settings.schema_json(indent=4, ensure_ascii=False))
```
<details>

```json
{
    "title": "Settings",
    "type": "object",
    "properties": {
        "service": {
            "$ref": "#/definitions/Service"
        }
    },
    "required": [
        "service"
    ],
    "definitions": {
        "Service": {
            "title": "Service",
            "type": "object",
            "properties": {
                "urls": {
                    "title": "Urls",
                    "default": {
                        "ac762f77-1e77-4ff8-906a-fd63ca622938": "https://example.com"
                    },
                    "type": "object",
                    "additionalProperties": {
                        "type": "string"
                    }
                }
            }
        }
    }
}
```

</details>

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: .venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.1 (main, Jan 31 2022, 09:16:52) [Clang 13.0.0 (clang-1300.0.29.3)]
                     platform: macOS-12.4-arm64-arm-64bit
     optional deps. installed: ['dotenv', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [X] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by Bobronium**
Closes #4774

Fix was pretty straightforward, I guess we can have this in 1.10.x.

### CODE REVIEW

**Comment by Bobronium** [comment]
CI seems to be failing with installation error unrelated to the changes.

**Comment by hramezani** [comment]
I created https://github.com/pydantic/pydantic/pull/4782 to fix the CI problem.
I am not sure about the solution

**Comment by Bobronium** [comment]
Perhaps I can rebase `Bobronium:1.10.X-fix-schema-model-default` on `hramezani:setuptools`?

**Comment by hramezani** [comment]
@Bobronium I would suggest waiting for @samuelcolvin to review and probably merge my PR. then you can rebase on top of main

**Comment by hramezani** [comment]
FYI, I closed https://github.com/pydantic/pydantic/pull/4782 because it was on main and seems not a proper fix

**Comment by hramezani** [comment]
As the problem got fixed in https://github.com/pydantic/pydantic/commit/78ca710fe769552d91c12adf972a9c88a2a1a90f, you can rebase your patch now!

**Comment by Bobronium** [comment]
Please review :)

File: tests/test_schema.py [hunk 1]
```diff
@@ -1523,6 +1523,39 @@ class UserModel(BaseModel):
     }
 
 
+def test_model_default():
+    """Make sure inner model types are encoded properly"""
+    from pydantic import BaseModel
+
```
Comment (hramezani) [inline tests/test_schema.py:1529]: I think we don't need this import here
```suggestion
```

File: pydantic/schema.py [hunk 2]
```diff
@@ -971,15 +972,20 @@ def multitypes_literal_field_for_schema(values: Tuple[Any, ...], field: ModelFie
...
         return dft.value
     elif isinstance(dft, (int, float, str)):
         return dft
     elif isinstance(dft, (list, tuple)):
         t = dft.__class__
         seq_args = (encode_default(v) for v in dft)
         return t(*seq_args) if is_namedtuple(t) else t(seq_args)
-    elif isinstance(dft, dict):
+
+    if isinstance(dft, dict):
```
Comment (hramezani) [inline pydantic/schema.py:988]: Do we need this change here?

**Comment by hramezani** [comment]
@Bobronium I left a suggestion and asked a question.

Please update

File: pydantic/schema.py (hunk 2 above)
Comment (Bobronium) [reply pydantic/schema.py:988]: Yes. Notice the change above: we transform models and dataclasses to `dict` in first `if` and need this to be a separate one to work.

**Comment by Bobronium** [comment]
Nice catch with suggestion, @hramezani! I answered the question as well.

Please review.

File: pydantic/schema.py (hunk 2 above)
Comment (Bobronium) [reply pydantic/schema.py:988]: Any ideas how to make this more explicit and obvious, btw?

Comment (samuelcolvin) [reply pydantic/schema.py:988]: maybe just a comment with the above explanation.

**Review by samuelcolvin — COMMENTED** [review]
Otherwise LGTM.

File: pydantic/schema.py (hunk 2 above)
Comment (hramezani) [reply pydantic/schema.py:988]: I would add the first if as a separate if and move the `isinstance(dft, dict)` check to the top.

Here is my suggestion(I haven't tested it locally):
```python
def encode_default(dft: Any) -> Any:
    from .main import BaseModel

    if isinstance(dft, BaseModel) or is_dataclass(dft):
        dft = cast('dict[str, Any]', pydantic_encoder(dft))

    if isinstance(dft, dict):
        return {encode_default(k): encode_default(v) for k, v in dft.items()}
    elif isinstance(dft, Enum):
        return dft.value
    elif isinstance(dft, (int, float, str)):
        return dft
    elif isinstance(dft, (list, tuple)):
        t = dft.__class__
        seq_args = (encode_default(v) for v in dft)
        return t(*seq_args) if is_namedtuple(t) else t(seq_args)
    elif dft is None:
        return None
    else:
        return pydantic_encoder(dft)

```

Comment (Bobronium) [reply pydantic/schema.py:988]: Happy to accept suggestions on phrasing as well :)
```suggestion
    # we can end up here either from first condition or if none of above conditions were met
    if isinstance(dft, dict):
```

Comment (Bobronium) [reply pydantic/schema.py:988]: Thanks for suggestion, @hramezani, I like your approach! I think I was hesitant on reordering at first, but it seems like the best move here, indeed.

**Review by hramezani — APPROVED** [review]
LGTM :+1: 
Thanks @Bobronium

**Comment by samuelcolvin** [comment]
Thanks so much.

I've also cherry-picked this into `main`, mostly so we don't forget this edge case when working on JSONSchema for V2.

**Pull request merged**